### PR TITLE
Fix setting a new STFC PI who hasn't been prev collaborated with

### PR DESCRIPTION
**Pull request by simonfernandes**
## Description

When choosing a new PI who isn't in the users recently collaborated list, the field appeared blank as if it hadn't worked, which was misleading. It now displays the name correctly. The bug was caused by not being able to fetch the user details due to them not being associated with the proposal. 

Part of UserOfficeProject/user-office-project-issue-tracker#856
## Motivation and Context

This issue was blocking this PR to allow quick switching between Co-I and PI: https://github.com/UserOfficeProject/user-office-core/pull/236

## How Has This Been Tested

Manual testing by adding a PI and seeing them appear, and then submitting the proposal.

## Fixes

Fixes UserOfficeProject/user-office-project-issue-tracker#856

## Changes

STFC specific files

## Tests included/Docs Updated?

- [ ] I have added tests to cover my changes.
- [ ] All relevant doc has been updated

### CODE REVIEW

File: apps/user-office-backend/src/auth/StfcUserAuthorization.ts [hunk 1]
```diff
@@ -44,6 +44,10 @@ export class StfcUserAuthorization extends UserAuthorization {
     Tokens.InstrumentDataSource
   );
 
+  protected userDataSource: StfcUserDataSource = container.resolve(
+    Tokens.UserDataSource
+  ) as StfcUserDataSource;
```
Comment (simonfernandes) [inline apps/user-office-backend/src/auth/StfcUserAuthorization.ts:49]: I don't think it's the best practice to use type assertion here, but I opted to do this over adding the `isSearchableUser` function to the `UserDataSource` interface. I think in this case we can always reasonably expect that the STFC auth would use the STFC data source, but let me know if the other way is better practice.

**Review by bolmsten — APPROVED** [review]

File: apps/user-office-backend/src/auth/StfcUserAuthorization.ts (hunk 1 above)
Comment (mutambaraf) [reply apps/user-office-backend/src/auth/StfcUserAuthorization.ts:49]: Looks good to me

File: apps/user-office-backend/src/auth/StfcUserAuthorization.ts [hunk 2]
```diff
@@ -303,4 +307,16 @@ export class StfcUserAuthorization extends UserAuthorization {
...
+  async canReadUser(agent: UserWithRole | null, id: number): Promise<boolean> {
+    const readableUsers = await this.listReadableUsers(agent, [id]);
+
+    if (readableUsers.includes(id)) {
+      return true;
+    }
+
+    const isSearchableUser = await this.userDataSource.isSearchableUser(id);
+
+    return isSearchableUser;
```
Comment (mutambaraf) [inline apps/user-office-backend/src/auth/StfcUserAuthorization.ts:320]: We can return readableUsers.includes(id) || await this.userDataSource.isSearchableUser(id) if possible.

File: apps/user-office-backend/src/datasources/stfc/StfcUserDataSource.ts [hunk 3]
```diff
@@ -518,7 +518,12 @@ export class StfcUserDataSource implements UserDataSource {
   async getRoleByShortCode(roleShortCode: Roles): Promise<Role> {
     throw new Error('Method not implemented.');
   }
+
   mergeUsers(fromUserId: number, intoUserId: number): Promise<void> {
     throw new Error('Method not implemented.');
   }
+
+  async isSearchableUser(userId: number): Promise<boolean> {
+    return !!(await this.getStfcBasicPersonByUserNumber(String(userId), true));
```
Comment (mutambaraf) [inline apps/user-office-backend/src/datasources/stfc/StfcUserDataSource.ts:527]: We can do userid.toString() here if possible.

**Review by mutambaraf — COMMENTED** [review]
Looking good just minor suggestions happy to approve.

**Pull request merged**